### DISCUSSION TITLE
How to parametrize a deployment?

### DISCUSSION THREAD

**Discussion by sfermigier** [pyinfra-dev/pyinfra#1034] [Q&A]
I'm trying to parametrize a deployment script for an application. I want a "dev" deployment (-> vagrant) and a "prod" deployment (-> production server), possibly with extra parameters (e.g. app name and ports, for multi-instance deployments).

So far I have found 2 options:

- Using an env var, e.g. `PROFILE=dev pyinfra @vagrant deploy.py`, with `deploy.py` interpreting the env variable to use the intended parameters (which are hardcoded in the script).
- Having several deploy scripts (e.g. `deploy-dev.py`, `deploy-prod1.py`, etc.) that contain the proper parameters and call a common set of functions contained in another module (I called it `recipes.py`).

Is this a common use case? Is there a recommended solution? Did I miss a way to pass a parameter or a config to the `pyinfra` CLI?

**Comment by maisim** [comment]
@sfermigier 

Not sure to well understand but it seems better to use hosts groups :
```python
# in inventory.py:
dev  = ["srvdev1"]
prod = ["srvprod1", "srvprod2"]

# then in your deploy.py:
if "prod" in host.groups:
    ...
```

You can also use host data:
```python
# in inventory.py:
my_app_servers = [
    ("srvprod1", {"run_env": "prod"}),
    ("srvprod2", {"run_env": "prod"}),
    ("srvdev1", {"run_env": "dev"}),
]

# then in your deploy.py:

if host.data.run_env == "prod":
    ...
elif host.data.run_env == "dev":
    ...
else:
    raise DeployError("Unknown run_env")
```
